<?xml version="1.0"?>
<launch>
    <!-- For controller test -->

    <arg name="world" default="worlds/empty.world"/>
    <arg name="x" default="0.0"/>
    <arg name="y" default="0.0"/>
    <arg name="z" default="0.4"/>
    <arg name="R" default="0"/>
    <arg name="P" default="0"/>
    <arg name="Y" default="0"/>

    <!-- vehicle model and world -->
    <arg name="est" default="ekf2"/>
    <arg name="vehicle" default="typhoon_h480"/>
    <arg name="sensor" default="typhoon_h480_kepco"/>
    <arg name="urdf" default="$(find kepco_sim)/gazebo_env/drone_models/$(arg sensor)/$(arg sensor).urdf"/>
    <!-- <arg name="urdf" default="/home/seunghyun/seunghyun/aerial_gym_simulator/resources/robots/quad/model.urdf"/> -->

    <!-- GT to vision pose (URDF cannot use GPS, SDF cannot use manipulator...) -->
    <param name ="/drone_name" value="$(arg vehicle)"/>
    <param name ="/model_name" value="$(arg sensor)"/>
    <node name="gt_vision" pkg="kepco_sim" type="gt_vision_pose.py"/>
    
    <!-- Wind disturbance -->    
    <rosparam file="$(find kepco_sim)/config/test_param.yaml" command="load"/>  
    <node name="wind_disturbance" pkg="kepco_sim" type="kepco_sim_wind_node"/>
    
    <!-- Manual controller with arm -->
    <node name="manipulator_controller" pkg="kepco_sim" type="manipulator_controller.py"/>

    <!-- UAV path visualize-->
    <node name="path_publisher" pkg="kepco_sim" type="path_publisher.py" output="screen"/>

    <!-- Do not touch this block -->
    <arg name="description_name" default="robot_description"/>
    <param name="$(arg description_name)" command="$(find xacro)/xacro --inorder '$(arg urdf)'"/>
    <rosparam file="$(find kepco_sim)/gazebo_env/drone_models/$(arg sensor)/ros_control.yaml" command="load"/>
    <!-- <node ns="$(arg sensor)" name="controller_spawner" pkg="controller_manager" type="spawner" respawn="false" args="joint_group_position_controller"/> -->

    <!-- gazebo configs -->
    <arg name="gui" default="false"/>
    <arg name="debug" default="false"/>
    <arg name="verbose" default="true"/>
    <arg name="paused" default="false"/>
    <arg name="respawn_gazebo" default="false"/>
    <!-- MAVROS configs -->
    <arg name="fcu_url" default="udp://:14540@localhost:14557"/>
    <arg name="respawn_mavros" default="false"/>
    <!-- PX4 configs -->
    <arg name="interactive" default="true"/>
    <!-- PX4 SITL and Gazebo -->
    <include file="$(find kepco_sim)/launch/posix_sitl_urdf.launch"> <!-- important -->
        <arg name="x" value="$(arg x)"/>
        <arg name="y" value="$(arg y)"/>
        <arg name="z" value="$(arg z)"/>
        <arg name="R" value="$(arg R)"/>
        <arg name="P" value="$(arg P)"/>
        <arg name="Y" value="$(arg Y)"/>
        <arg name="world" value="$(arg world)"/>
        <arg name="vehicle" value="$(arg vehicle)"/>
        <arg name="robot_description" value="$(arg description_name)"/>
        <arg name="gui" value="$(arg gui)"/>
        <arg name="interactive" value="$(arg interactive)"/>
        <arg name="debug" value="$(arg debug)"/>
        <arg name="verbose" value="$(arg verbose)"/>
        <arg name="paused" value="$(arg paused)"/>
        <arg name="respawn_gazebo" value="$(arg respawn_gazebo)"/>
    </include>
    <!-- MAVROS -->
    <include file="$(find mavros)/launch/px4.launch">
        <!-- GCS link is provided by SITL -->
        <arg name="gcs_url" value=""/>
        <arg name="fcu_url" value="$(arg fcu_url)"/>
        <arg name="respawn_mavros" value="$(arg respawn_mavros)"/>
    </include>

    <node type="rviz" name="rviz" pkg="rviz" args="-d $(find sh_mpc)/test.rviz" />
</launch>
